test/sharness/t0110-gateway.sh

```diff
@@ -22,58 +22,315 @@ apiport=$PORT_API
+  test_cmp expected actual || test_fsh cat actual.headers
+}
+
+immutable_ttl=$((10*365*24*60*60))
+unknown_ttl=60
+
+# DATA1, HASH1: a file
+# DATA2, HASH2: a directory
+# PEERID: the node ID
+test_expect_success "Generating test environment" '
+  DATA1=data1 &&
+  echo "Hello Worlds!" >"$DATA1" &&
+  HASH1=$(ipfs add -q "$DATA1") &&
+  DATA2=data2 &&
+  mkdir "$DATA2" &&
+  echo "12345" >"$DATA2"/test &&
+  mkdir "$DATA2"/has-index-html &&
+  echo "index" >"$DATA2"/has-index-html/index.html &&
+  ipfs add -r -q "$DATA2" >actual &&
+  HASH2=$(tail -n 1 actual) &&
+  PEERID=$(ipfs config Identity.PeerID) &&
+  test_check_peerid "$PEERID"
 '
 
-test_expect_success "GET IPFS path output looks good" '
-  test_cmp expected actual &&
-  rm actual
+ITEM="GET IPFS path"
+
+test_expect_success "$ITEM: succeeds" '
+  rm -f actual.headers actual &&
+  test_curl_headers actual.headers actual "http://127.0.0.1:$port/ipfs/$HASH1" &&
+  test_should_contain "^HTTP/1.1 200 OK.\$" actual.headers
 '
 
-test_expect_success "GET IPFS path on API forbidden" '
-  test_curl_resp_http_code "http://127.0.0.1:$apiport/ipfs/$HASH" "HTTP/1.1 403 Forbidden"
+test_expect_success "$ITEM: output looks good" '
+  test_cmp "$DATA1" actual
 '
 
-test_expect_success "GET IPFS directory path succeeds" '
-  mkdir dir &&
-  echo "12345" >dir/test &&
-  ipfs add -r -q dir >actual &&
-  HASH2=$(tail -n 1 actual) &&
-  curl -sf "http://127.0.0.1:$port/ipfs/$HASH2"
+test_expect_success "$ITEM: has long Cache-Control max-age" '
+  test_expect_max_age actual.headers "$immutable_ttl"
 '
 
-test_expect_success "GET IPFS directory file succeeds" '
-  curl -sfo actual "http://127.0.0.1:$port/ipfs/$HASH2/test"
+# The go http package replaces ETag with Etag.  Accept both.  (HTTP header
+# names are case-insensitive but ETag is the official name.)
+test_expect_success "$ITEM: has ETag" '
+  test_should_contain "^E[Tt]ag: \"$HASH1\".\$" actual.headers
 '
 
-test_expect_success "GET IPFS directory file output looks good" '
-  test_cmp dir/test actual
+ITEM="GET IPFS path on API"
+
+test_expect_success "$ITEM: forbidden (403)" '
+  rm -f actual.headers actual &&
+  test_curl_headers actual.headers actual "http://127.0.0.1:$apiport/ipfs/$HASH1" &&
+  test_should_contain "^HTTP/1.1 403 Forbidden.\$" actual.headers
 '
 
-test_expect_success "GET IPFS non existent file returns code expected (404)" '
-  test_curl_resp_http_code "http://127.0.0.1:$port/ipfs/$HASH2/pleaseDontAddMe" "HTTP/1.1 404 Not Found"
+test_expect_success "$ITEM: has no Cache-Control max-age" '
+  test_expect_no_max_age actual.headers
 '
 
-test_expect_failure "GET IPNS path succeeds" '
-  ipfs name publish "$HASH" &&
-  PEERID=$(ipfs config Identity.PeerID) &&
-  test_check_peerid "$PEERID" &&
-  curl -sfo actual "http://127.0.0.1:$port/ipns/$PEERID"
+test_expect_success "$ITEM: has no ETag" '
+  test_should_not_contain "^ETag: " actual.headers
+'
+
+ITEM="GET IPFS directory path"
+
+test_expect_success "$ITEM: succeeds" '
```

Comment:
They are different. This one checks for `test_should_contain "Index of /ipfs/$HASH2" actual` while the following one checks `test_cmp "$DATA2"/test actual`. The one after that checks that a redirect happens. The one after that checks a path which contains `index.html` has to check the ETag header against the path that was requested, not against the hash of the `index.html`.

It gets even more difficult to refactor them to a function later where we want to do an IPNS publish for the following item but check that the cache still returns results from the earlier IPNS publish for the current item.

This is an annoying case where all the items are just slightly different in various different ways and sh doesn’t exactly provide powerful ways to abstract things.

But I’ll give it another look and see if I can find a way to make it a bit better.